src/vivarium_census_prl_synth_pop/data/generate_business_names.py

```diff
@@ -0,0 +1,148 @@
+
+    business_names = pd.read_csv(paths.BUSINESS_NAMES_DATA)
+    bigrams = make_bigrams(business_names)
+
+    # Get frequency of business names and find uncommon ones
+    s_name_freq = business_names.location_name.value_counts()
+    real_but_uncommon_names = set(s_name_freq[s_name_freq < 1_000].index)
+
+    # Generate random business names.  Drop duplicates and overlapping names with uncommon names
+    n_total_names = int(n_total_names)  # Make int because of sys args
+    new_names = pd.Series()
+
+    # Generate additional names until desired number of random business names is met
+    while len(new_names) < n_total_names:
+        n_needed = n_total_names - len(new_names)
+        more_names = sample_names(
+            bigrams, n_needed, data_values.BUSINESS_NAMES_MAX_TOKENS_LENGTH
+        )
+        new_names = pd.concat([new_names, more_names]).drop_duplicates()
+        new_names = new_names.loc[~new_names.isin(real_but_uncommon_names)]
+
+    new_names.to_csv(
+        paths.BUSINESS_NAMES_DATA_ARTIFACT_INPUT_PATH,
+        header=["business_names"],
+        index=False,
+        compression="bz2"
+    )
+
+
+def make_bigrams(df: pd.DataFrame) -> defaultdict:
+    """
+
+    Parameters
+    ----------
+    df: A one column dataframe from a CSV containing a list of real business names.
+
+    Makes default dict of business names for map to sample from
+    Bigrams will be a Dict[str: Dict[str: int]]
+
+    Example:  A series of business names to be turned into a Dict map
+            location_name
+        0 	Cottage Inn Pizza
+        1 	Union St Barber Shop
+        2 	Larkfield Auto Center
+        3 	Randall Guns
+        4 	Sport Clips
+
+    Create bigrams to map a word and the frequence at which the next choice happens.
+
+        {'<start>': {'Cottage': 1,
+                     'Union': 1,
+                     'Larkfield': 1,
+                     'Randall': 1,
+                     'Sport': 1
+                     }
+        }
+
+    Go through each word in each name and get the frequency of each name pair in the series
+    Returns
+    -------
+    defaultdict
+    """
+
+    def dict_factory():
+        return lambda: defaultdict(int)
+
+    bigrams = defaultdict(dict_factory())
+    n_rows = len(df)  # expect a few minutes
+
+    for i in range(n_rows):
+        if i % 10_000 == 0:
+            print(".", end=" ", flush=True)
+        names_i = df.iloc[i, 0]
+
+        tokens_i = names_i.split(" ")
+        for j in range(len(tokens_i)):
+            if j == 0:
+                bigrams["<start>"][tokens_i[j]] += 1
+            else:
+                bigrams[tokens_i[j - 1]][tokens_i[j]] += 1
+        bigrams[tokens_i[j]]["<end>"] += 1
+
+    return bigrams
+
+
+def sample_names(bigrams: defaultdict, n_businesses: int, n_max_tokens: int) -> pd.Series:
+    """
+
+    Parameters
+    ----------
+    bigrams: Default dict produced from make_bigrams function (see formatting of default dict.
```

Comment:
could you update the naming or the comments so it's clearer what this `bigrams` contains?

Reply:
I've updated the docstring.  Im trying to keep the names for the different data objects throughout the module so it is (hopefully) easier to follow.